### ISSUE TITLE
ESM export?

### ISSUE THREAD

**Issue by vlaraort** [guidepup/virtual-screen-reader#49]
# Issue

I am trying to try this awesome library in my project, but I just found that it cannot be imported as a ESM module from my test files.

I am using web-test-runner (https://modern-web.dev/docs/test-runner/writing-tests/js-tests/) to run my tests, and as you can see, ESM imports are necesary, but it doesn't support TS or require/AMD exports

Would be possible to build the project allowing to do imports from JS files, like it's in TS?

Thanks!

**Comment by vlaraort** [comment]
I have tried to build the project for esnext but I have some problems into some dependencies. If you want I can provide a sample project to debug, logs or what you need :)

**Comment by cmorten** [comment]
Hey @vlaraort 👋 

The ask is straight-forward enough so should be reasonable to do.

Some thought needed as to the best approach as there are a few approaches, e.g. using .cjs and .mjs vs 2 package.json alternatives.

In the meantime you might be able to unblock yourself by using the likes of https://jspm.org/, https://esm.sh/, or https://jsr.io/

**Comment by vlaraort** [comment]
I don't think you need two package.json, only ship two builds in the same package, and use this approach

https://antfu.me/posts/publish-esm-and-cjs

I will check in the meantime the solutions you provide,  thanks for the quick response and the effort of this library! I think it's a awesome solution for having regression testing on accessibility, and it's something really needed in this times to aware about a11y ;)

**Comment by cmorten** [comment]
Yep that’s the first (and newer) approach where you ship different extensions and utilise the newish exports field 😊

**Comment by cmorten** [comment]
Hi @vlaraort 👋 

I’ve hopefully now updated the package to dual support CJS and ESM. Please could you try with `0.18.0`?

**Comment by vlaraort** [comment]
I tried, thanks!

![image](https://github.com/guidepup/virtual-screen-reader/assets/2410951/8371d229-0674-4c76-aac5-3a367c89d2a5)

I received the following error and it crashes, I am going to check that dependency to see if this can be fixed :)

**Comment by vlaraort** [comment]
In fact, here are some information regarding the same dependency, in the same library 

https://github.com/A11yance/aria-query/issues/48

**Comment by cmorten** [comment]
> I tried, thanks!
> 
> ![image](https://private-user-images.githubusercontent.com/2410951/306191519-8371d229-0674-4c76-aac5-3a367c89d2a5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg0MTk3OTEsIm5iZiI6MTcwODQxOTQ5MSwicGF0aCI6Ii8yNDEwOTUxLzMwNjE5MTUxOS04MzcxZDIyOS0wNjc0LTRjNzYtYWFjNS0zYTM2N2M4OWQyYTUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIyMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMjBUMDg1ODExWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NTNiNjIwMmFkNjA1YTkzM2QwMjhjZjFhZjI1Y2IwOWJmYmFhZmUxOWJlYzBkNmExZDFhOTlmOGNkMTZkODlmOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.x--T26UE2u_vjtSPvf0kLO5a8Q-KSXYRZ0LV9LPa-YY)
> 
> I received the following error and it crashes, I am going to check that dependency to see if this can be fixed :)

Ah apologies, I should have tested more thoroughly 🤦 

I’ll take a look as to whether there is anything I can do in this package to work around the issue, otherwise I guess we ask the question again on the aria-query thread!

**Comment by vlaraort** [comment]
Looks like they don't really want to provide a ESM export because reasons

https://github.com/A11yance/aria-query/pull/485

But, there is a fork that makes the ESM build, maybe we can try with this fork of aria-roles instead?

https://github.com/KonnorRogers/aria-query/tree/konnorrogers/publish-modern-build

**Comment by vlaraort** [comment]
I have been playing around with the dependencies, making overrides from aria-roles to aria-roles-esm, and it's not the only dependency that causes problems in a ESM build. I also hit the wall with lz-string, that doesn't like that is going to work, so looks like it would be more difficult than assumed.

The only solution that comes to my mind (and I don't like too much), is to create a webpack/rollup/whatever bundle with all the dependencies embeed, so they are also processed and converted to ESM code. WDYT?

**Comment by cmorten** [comment]
See also https://github.com/testing-library/dom-testing-library/issues/995 it looks like this has been raised before, but the issue was quite muddled and closed based on a change that doesn’t quite go the whole way.

Maybe worth raising an issue on Testing Library to see what appetite would be?

In the meantime using the likes of https://esm.sh/@guidepup/virtual-screen-reader@0.19.0 as mentioned before might be the way to go for now

**Comment by cmorten** [comment]
See https://modern-web.dev/docs/dev-server/plugins/rollup/ for a guide on how to use the Rollup plugins within web test runner, specifically the commonjs one.

Away from desk so can’t test atm

**Comment by scottnath** [comment]
Hi @cmorten, hi @vlaraort,

I spent some time digging into this and I was able to get it running in an esm environment (a storybook story) by making the following changes. The aria-query problem can be solved by adjusting how installs work, but it would require maintaining and updating `aria-query-esm`. 

The guidepup problem would require some thought on how to adjust WindowsModifiers so it doesn't rely on `fs/promises` (not sure if other node script imports are problems too). Since VSR is system-agnostic, I guess I'd want to know what the os-specific modifiers are needed for in VSR?

## aria-query

To replace aria-query, I changed package.json on VSR to the below. The `resolutions` will replace all nested uses of aria-query

```
...

    //"aria-query": "^5.3.0", <- replaced with aria-query-esm
    "aria-query-esm": "^5.3.0-fix.1",
    "dom-accessibility-api": "^0.6.3"
  },
  "devDependencies": { 
    ...
  },
  // this supersedes all dependency uses to instead use *-esm
  "resolutions": {
    "aria-query": "https://registry.yarnpkg.com/aria-query-esm/-/aria-query-esm-5.3.0-fix.1.tgz"
  }
}
```

## Turn off WindowsModifiers from guidepup

The next error I kept seeing was coming from guidepup. This solution is not a workable one because it would turn off features, but it does help narrow down where the problem lies. MacOSModifiers are not causing issues

### bug

```
node_modules/@guidepup/guidepup/lib/windows/runVbsScript.js:5:27:
      5 │ const promises_1 = require("fs/promises");
```

### hacky solution

In `src/Virtual.ts` I turned off 

```
// turned off importing WindowsModifiers
import {
  CommandOptions,
  MacOSModifiers,
  ScreenReader,
  // WindowsModifiers,
} from "@guidepup/guidepup";

// line 763, removed the `WindowsModifiers` ref here
    rawKeys.forEach((rawKey) => {
      if (
        typeof MacOSModifiers[rawKey] !== "undefined"
      ) {
        modifiers.push(rawKey);
      } else {
        keys.push(rawKey);
      }
    });
```

**Comment by scottnath** [comment]
Addendum:

Looks like this can be fixed with a patch-level change in guidepup. I detailed the problem here:
https://github.com/guidepup/guidepup/issues/77

**Comment by jlp-craigmorten** [comment]
Thanks @scottnath for diving in!

Are you able to share a simple example of this working for storybook, would be great to have a reference as well as get an example into this repo? (A contribution would be awesome but no pressure there!)

This doesn't feel too different to a wip branch had on the go to see if could resolve this https://github.com/guidepup/virtual-screen-reader/pull/56, though it still wasn't working - will revisit with your comments and see if there's anything obvious!

**Comment by scottnath** [comment]
@jlp-craigmorten after the https://github.com/guidepup/guidepup/pull/78 changes are in, I can def push my storybook example where I'm experimenting with VSR in [my profile-components library](https://github.com/scottnath/profile-components)

FWIW - fixing guidepup fixed the errors I was seeing. I was able to use VSR 0.19.0 as-is with the guidepup changes without changing aria-query. My setup may be different than the og of this thread tho I realize.

**Comment by jlp-craigmorten** [comment]
@vlaraort please see https://github.com/guidepup/virtual-screen-reader/tree/main/examples/web-test-runner for an example of how to get this package working with web-test-runner.

Due to upstream deps being written in CJS which is out of this project's control it requires some bundling steps to work. It is often frowned upon to have packages bundle themselves so that such concerns can be left to the consumer - appreciate that doesn't make this an out of the box solution for you, but hopefully there isn't too much overhead here.

@scottnath the new `0.20.0` version also has the `@guidepup/guidepup` changes.

**Comment by scottnath** [comment]
@jlp-craigmorten I've got that working storybook!

Your PR made a massive difference. I had a whole thing that was using a Vite plugin to mimic Node and another esbuild to change an node_modules deep file. But removing guidepup/guidepup let me get rid of all that.

Here's a working storybook:

GitHub: https://github.com/scottnath/virtual-screen-reader-storybook
StackBlitz dev env: https://stackblitz.com/~/github.com/scottnath/virtual-screen-reader-storybook

Excellent change - this will make it SUPER easy to use VSR in storybook. Thanks!

**Comment by jlp-craigmorten** [comment]
Going to close this as we have demonstrated this package working in a couple of ESM settings.

I appreciate it still requires a bundler of some form on the consumers end, but to resolve that would require work upstream to Testing Library and aria-query. Issues can/should be raised against those repositories to resolve the root issue, and if/once fixed there, we can update this package.

**Comment by jlp-craigmorten** [comment]
By way of update, as of `0.22.1` this package now exports a standalone browser bundle that is ESM, minified, and includes third party scripts that can be used in Web Test Runner and other ESM only environments without a build step.

To import use:

```ts
import { virtual } from '@guidepup/virtual-screen-reader/browser.js';
```

Or for dynamic imports:

```ts
const { virtual } = await import('@guidepup/virtual-screen-reader/browser.js');
```

Or if using a CDN (e.g. directly in a browser):

```ts
const { virtual } = await import("https://unpkg.com/@guidepup/virtual-screen-reader");
```

**Comment by philsherry** [comment]
Craig, I'll add this here for now rather than as a new Issue, because it seems related _enough_ to tie it into this one for now. 

I’ve been wrestling with trying to get this working, but I just cannot get a result in the project I need it working in. I can spin up a small demo project and get it working in there, but when I try to add it into the big TS-driven project (that I can’t share code or screenshots from—government thing)‚ I get one of the following errors in my console. These are all after trying what you suggested in the comment above. 

```sh
ReferenceError: expect is not defined
```

```sh
error TS2339: Property 'toMatchScreenReaderSnapshot' does not exist on type 'JestMatchers<HTMLElement>'.
```

```sh
error TS2339: Property 'toMatchScreenReaderInlineSnapshot' does not exist on type 'JestMatchers<HTMLElement>'.
```

```sh
SyntaxError: Unexpected token 'export'

> 1 | import '@guidepup/jest';
    | ^
```

```sh
error TS7016: Could not find a declaration file for module '@guidepup/virtual-screen-reader/browser.js'. '…/node_modules/@guidepup/virtual-screen-reader/lib/esm/index.browser.js' implicitly has an 'any' type.
    Try `npm i --save-dev @types/guidepup__virtual-screen-reader` if it exists or add a new declaration (.d.ts) file containing `declare module '@guidepup/virtual-screen-reader/browser.js';`
```

Any thoughts or guidance would be great, because I’m at desk-tipping stage here. 

Cheers.

**Comment by cmorten** [comment]
Hey @philsherry 

Going by the above errors I’m guessing you’re attempting to use https://github.com/guidepup/jest opposed to this package directly?

Given this issue was originally related to running with web-test-runner, for completeness (although sure you have this grok’d) in order to use the Jest matchers you’ll first have to install and configure Jest for your environment before expecting the `@guidepup/jest` package to work.

The `expect` error you’ve listed is usually a sign that something about your Jest setup might well be off - `expect` should be globally available. Doing a quick search for this it looks like the README might be off for that repo, see https://github.com/enzymejs/enzyme-matchers/issues/86#issuecomment-312489052

> If you want to extend expect, you must do it in the file specified by setupTestFrameworkScriptFile in your jest config and not setupFiles.

Worth trying adjusting the config and seeing if resolves that?

The following two errors might then resolve owing to the package being consumed ok.

R.E. the unexpected export token error, this might be because the jest package hasn’t had any ESM support consideration - if using in that fashion we might need to open an issue there and do some work!

For the last error it looks like might need to fix up this packages declaration files and package entries, looks like it’s not quite finding it.

—-

If you are able to figure out how to repo on a small example project that would be awesome for triaging and debugging! Appreciate that you’ve tried already - will have a go my end.

**Comment by philsherry** [comment]
Hellish weekend, bled over into the week, so I’m only just getting back to this now. 😬 

Okay:

- trying to setup both `@guidepup/jest` and `@guidepup/virtual-screen-reader` in the project
- project already has `import { expect, jest, test } from '@jest/globals';` in `jest.setup.js`
- `const { virtual } = await import('@guidepup/virtual-screen-reader');` works within an individual `test()` block, but it would obviously be much better to include as `import('@guidepup/virtual-screen-reader')` at the top of the file—or even in the `jest.setup.js` file

In case this helps, here’s some visual feedback. Note the wall of minified JS with an `export` at the end—that’s what it’s complaining about.

<img width="2160" alt="guidepup-error-1" src="https://github.com/guidepup/virtual-screen-reader/assets/272769/6cfc294d-d2fe-4370-8581-b3facd6021d1">

<img width="2560" alt="guidepup-error-2" src="https://github.com/guidepup/virtual-screen-reader/assets/272769/567651e7-9ba7-4452-beae-33917b4d9870">

**Comment by cmorten** [comment]
Can you try importing `@guidepup/jest` into a jest setup file that is listed against `setupFilesAfterEnv` instead of `setupFiles`, i.e.

```ts
  setupFiles: [ /* your current / normal setup files */],
  setupFilesAfterEnv: [ "<rootDir>/setup-guidepup-jest.js" ],
```

(Apologies, in comment above I referenced `setupTestFrameworkScriptFile` which is an outdated key)

This should hopefully resolve the errors around `expect` not be defined.

(need to improve the readme and maybe could put in a more helpful error message!)

**Comment by philsherry** [comment]
Yeah, it’s already in the right key: `setupFilesAfterEnv: ['<rootDir>/jest.setup.js']`

Jest’s docs have had that right for ages.

**Comment by jlp-craigmorten** [comment]
Heya @philsherry 

Bare with, have reproduced and think making ground.

TIL a `testEnvironment` of `jsdom` will automatically assume a package export of `browser` which doesn't seem unreasonable... except ESM can be executed in a browser, but not handled at all well in a Jest context

REF: https://jestjs.io/docs/configuration#testenvironmentoptions-object

Going to try releasing a new version with some exports tweaks to get this workable, hopefully without additional config.

**Comment by jlp-craigmorten** [comment]
@philsherry can you try `@guidepup/virtual-screen-reader@^0.25.0` and `@guidepup/jest@^0.4.0` and see if it gets us any closer?

**Comment by philsherry** [comment]
Mate, looking good. `@guidepup/jest` is firing nicely from the setup file now. 

I’m seeing this, though, and feel free to take this up on the `@guidepup/jest` repo. 

```js
Property 'toMatchScreenReaderSnapshot' does not exist on type 'JestMatchers<HTMLElement>'.ts(2339)
```

Question about the inclusion of the file. I’m trying to have 2 × `describe` blocks for each page: 1 × `jest`, 1 × `virtual-screen-reader`. Prior to the change you just made, what you see below was the only way I could get it working. Am I missing something completely obvious; is there an easier way to include this? I think I’m at that stage of the day where I’d fail spelling my own name correctly.

```js
function setupA11yPage() {
  document.body.innerHTML = `
<div>
  …
</div>
`;
}

describe('matchers', () => {
  beforeEach(() => {
    setupA11yPage();
  });

  afterEach(() => {
    document.body.innerHTML = ``;
  });

  // I’m writing `@guidepup/jest` tests here — see above, where errors are mentioned.
  describe('snapshot tests', () => {
    test('…', async () => {});
  });

  // I’m writing `@guidepup/virtual-screen-reader` tests here.
  describe('virtual screen reader tests', () => {
    // navigate by headings
    test('…', async () => {
      // I can’t async the `describe` block, so I have to put this inside each
      // `test` block I write — which seems less than ideal, or wrong (to me).
      const { virtual } = await import('@guidepup/virtual-screen-reader');
    });
  });
});
```

**Comment by cmorten** [comment]
On the first point, let’s raise the issue on that repo and will take a look why the types aren’t propagating through.

On the latter, hopefully you should just be able use a single import statement at the top of the file:

```ts
import { virtual, Virtual } from "@guidepup/virtual-screen-reader";

function setupA11yPage() {
  document.body.innerHTML = `
<div>
  …
</div>
`;
}

describe('matchers', () => {
  beforeEach(() => {
    setupA11yPage();
  });

  afterEach(() => {
    document.body.innerHTML = ``;
  });

  describe('snapshot tests', () => {
    test('…', async () => {});
  });

  describe('virtual screen reader tests', () => {
    // navigate by headings
    test('…', async () => {
      // Use the easy start export
      await virtual.start({ … });
      // … rest of test

      // Or use the class constructor if you need multiple instances - e.g. sharding in parallel.
      const vsr = new Virtual();
      await vsr.start({ … });
      // … rest of test
    });
  });
});
```

Give package exports weren’t quite right before it may have prompted you async importing per test if was erroring?

**Comment by philsherry** [comment]
Even after making those changes, it’s barking at me in one of two ways. 

<img width="560" alt="" src="https://github.com/guidepup/virtual-screen-reader/assets/272769/024faede-ae5f-4500-b62a-79934cb677ad">

<img width="455" alt="" src="https://github.com/guidepup/virtual-screen-reader/assets/272769/c439bb99-85e5-4443-917c-1f7a3e6f2bcd">

Happy for you to open new Issues and just tag me into them.

**Comment by jlp-craigmorten** [comment]
@philsherry are you able to share the following?

- Are you using `"type": "module"` in your `package.json` or `.mjs`/`.mts` and `.cjs`/`.cts` extensions?
- What are the values you are using for these keys in your `tsconfig.json`? `target`, `module`, `moduleResolution`, and `esModuleInterop`

The first error indicates that your project is configured to consider the file a CommonJS module. The use of the import with VSR will target the ESM version of the library which can't be used in a CommonJS context _except_ as a dynamic import.

The second screenshot is how you import the bundled, minified, browser compatible ESM version of the library. This isn't something that would expect folks to use outside of a browser context.

Pending further info, would suggest your paths forward are one of:

1. Use a dynamic import at the beginning of your test files in place of a static import, and continue to use the CJS module setup as-is:

    ```ts
    const { Virtual } = await import("@guidepup/virtual-screen-reader");
    ```

2. Update your `package.json` and/or the file extension to indicate that this file is in fact ESM.
3. Update your `tsconfig.json` to use a `"moduleResolution": "bundler"` to allow module resolution via VSR's `exports` instead of through your package module type / extension.

**Comment by philsherry** [comment]
- Not using `"type": "module"` in our `package.json` or `.mjs/.mts` and `.cjs/.cts` extensions.
- Those `tsconfig.json` values:
```json
{
  "esModuleInterop": true,
  "module": "CommonJS",
  "moduleResolution": "node",
  "target": "es2020"
}
```

1. :white_check_mark: Will try this. 
2. :x: Cannot do this.
3. :x: Cannot do this. (Long story. TL;DR, our pipeline blows up.)

**Comment by cmorten** [comment]
Let me know how goes!

Failing that there is the option for the package to publish subpath exports to work around this, in your case something like `@guidepup/virtual-screen-reader/cjs`

Will have a play my end with an example matching that setup and see if any other options.

**Comment by philsherry** [comment]
Yeah, I get shouted at for having that at the top.

> `The current file is a CommonJS module and cannot use 'await' at the top level.ts(1309)`

**Comment by jlp-craigmorten** [comment]
@philsherry had another go fiddling with the package exports, can you try again with:

- `@guidepup/virtual-screen-reader@^0.26.0`
- `@guidepup/jest@^0.5.0`

**Comment by philsherry** [comment]
Updated. Now I’m getting this.
```sh
 FAIL  src/pages/__test__/AccessibilityPage.a11y.tsx
  ● Test suite failed to run

    src/pages/__test__/AccessibilityPage.a11y.tsx:4:25 - error TS1479: The current file is a CommonJS module whose imports will produce 'require' calls; however, the referenced file is an ECMAScript module and cannot be imported with 'require'. Consider writing a dynamic 'import("@guidepup/virtual-screen-reader")' call instead.
      To convert this file to an ECMAScript module, add the field `"type": "module"` to '…/package.json'.

    import { virtual } from '@guidepup/virtual-screen-reader';
```

FWIW, I’ve hated this project since I joined it.

**Comment by cmorten** [comment]
Hmm, starting to get a bit stuck then… without being able to see more of your projects setup 😅

Perhaps worth comparing against the https://github.com/guidepup/jest/blob/main/examples/typescript-legacy example and seeing where things are differing and sharing?

**Comment by philsherry** [comment]
I ended up being off sick for a couple of days. Came back to find your update edited in, so I've just updated things and… my tests are… passing? 🫣 

Note: Unless I add `@guidepup/jest` into the test file as well—despite it being referenced in the setup file—like so:
```js
import '@guidepup/jest';
import { virtual } from '@guidepup/virtual-screen-reader';
```

…then I get those errors I mentioned:
```sh
Property 'toMatchScreenReaderSnapshot' does not exist on type 'JestMatchers<null>'.
Property 'toMatchScreenReaderInlineSnapshot' does not exist on type 'JestMatchers<null>'.
```

But all is well if I add it into each test file. Not sure if that will mean anything to you, but I'm still happy to be tagged into any Issues for that particular beast.

Many thanks for sticking with me through this. 🙏
